3.0/m_relaymsg.cpp

```diff
@@ -0,0 +1,218 @@
+            invalid_chars_map.set(static_cast<unsigned char>(*n));
+        }
+    }
+
+    std::string GetFakeHostmask(std::string& nick) {
+        return InspIRCd::Format("%s!%s@%s", nick.c_str(), fake_ident.c_str(), fake_host.c_str());
+    }
+
+    CmdResult Handle(User* user, const CommandBase::Params& parameters)
+    {
+        std::string channame = parameters[0];
+        std::string nick = parameters[1];
+        std::string text = parameters[2];
+
+        // Check that the source has the relaymsg capability.
+        if (IS_LOCAL(user) && !cap.get(user)) {
+            return CMD_FAILURE;
+        }
+
+        Channel* channel = ServerInstance->FindChan(channame);
+        // Make sure the channel exists and the sender is in the channel
+        if (!channel)
+        {
+            user->WriteNumeric(Numerics::NoSuchChannel(channame));
+            return CMD_FAILURE;
+        }
+        if (!channel->HasUser(user))
+        {
+            user->WriteNumeric(Numerics::CannotSendTo(channel, "You must be in the channel to use this command."));
+            return CMD_FAILURE;
+        }
+
+        // Check that target nick is not already in use
+        if (ServerInstance->FindNick(nick))
+        {
+            user->WriteNumeric(ERR_BADRELAYNICK, nick, "RELAYMSG spoofed nick is already in use");
+            return CMD_FAILURE;
+        }
+
+        // Make sure the nick doesn't include any core IRC characters (e.g. *, !)
+        // This should still be more flexible than regular nick checking - in particular
+        // we want to allow "/" and "~" for relayers
+        for (std::string::const_iterator x = nick.begin(); x != nick.end(); x++)
+        {
+            if (invalid_chars_map.test(static_cast<unsigned char>(*x)))
+            {
+                user->WriteNumeric(ERR_BADRELAYNICK, nick, "Invalid characters in spoofed nick");
+                return CMD_FAILURE;
+            }
+        }
+
+        // If the sender was a lcoal user, check that the target includes a nick separator
+        if (IS_LOCAL(user))
+        {
+            bool accepted = false;
+            for (std::string::const_iterator x = cap.nick_separators.begin(); x != cap.nick_separators.end(); x++) {
+                if (nick.find(*x) != std::string::npos) {
+                    accepted = true;
+                }
+            }
+            if (!accepted)
+            {
+                user->WriteNumeric(ERR_BADRELAYNICK, nick, InspIRCd::Format("Spoofed nickname must include one of the following separators: %s", cap.nick_separators.c_str()));
+                return CMD_FAILURE;
+            }
+        }
+
+        // Send the message to everyone in the channel
+        std::string fakeSource = GetFakeHostmask(nick);
+        ClientProtocol::Messages::Privmsg privmsg(fakeSource, channel, text);
+        // Tag the message as @relaymsg=<nick> so the sender can recognize it
+        privmsg.AddTag("relaymsg", &captag, user->nick);
```

Comment:
You probably want to either draft prefix or vendor prefix this to avoid colliding with any potential ratified spec.

Reply:
Ooof, I totally forgot to add the prefix here too :/

It looks like the other implementations already use @relaymsg so this is a bit of a breaking change. I'll see if I can coordinate a fix?

Reply:
Fixed this here and in the spec https://github.com/ircv3/ircv3-specifications/pull/417/commits/70c0957bf53f5476d3b671f36ea9a4a0c086a5e